src/compiler/scala/tools/nsc/transform/async/AsyncAnalysis.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.tools.nsc.transform.async
+
+import scala.collection.mutable.ListBuffer
+import scala.reflect.internal.Flags
+
+trait AsyncAnalysis extends TransformUtils  {
+  import global._
+
+  /**
+   * Analyze the contents of an `async` block in order to:
+   * - Report unsupported `await` calls under nested templates, functions, by-name arguments.
+   *
+   * Must be called on the original tree, not on the ANF transformed tree.
+   */
+  def reportUnsupportedAwaits(tree: Tree): Unit = {
+    val analyzer = new UnsupportedAwaitAnalyzer
+    analyzer.traverse(tree)
+  }
+
+  private class UnsupportedAwaitAnalyzer extends AsyncTraverser {
+    override def nestedClass(classDef: ClassDef): Unit = {
+      val kind = if (classDef.symbol.isTrait) "trait" else "class"
+      reportUnsupportedAwait(classDef, s"nested $kind")
+    }
+
+    override def nestedModule(module: ModuleDef): Unit = {
+      reportUnsupportedAwait(module, "nested object")
+    }
+
+    override def nestedMethod(defDef: DefDef): Unit = {
+      reportUnsupportedAwait(defDef, "nested method")
+    }
+
+    override def byNameArgument(arg: Tree): Unit = {
+      reportUnsupportedAwait(arg, "by-name argument")
+    }
+
+    override def function(function: Function): Unit = {
+      reportUnsupportedAwait(function, "nested function")
+    }
+
+    override def patMatFunction(tree: Match): Unit = {
+      reportUnsupportedAwait(tree, "nested function")
+    }
+
+    override def traverse(tree: Tree): Unit = {
+      tree match {
+        case Try(_, _, _) if containsAwait(tree)              =>
+          reportUnsupportedAwait(tree, "try/catch")
+          super.traverse(tree)
+        case Return(_)                                        =>
+          global.reporter.error(tree.pos, "return is illegal within a async block")
+        case DefDef(mods, _, _, _, _, _) if mods.hasFlag(Flags.LAZY) && containsAwait(tree) =>
+          reportUnsupportedAwait(tree, "lazy val initializer")
+        case ValDef(mods, _, _, _) if mods.hasFlag(Flags.LAZY) && containsAwait(tree) =>
+          reportUnsupportedAwait(tree, "lazy val initializer")
+        case CaseDef(_, guard, _) if guard exists isAwait     =>
+          reportUnsupportedAwait(tree, "pattern guard")
+        case _                                                =>
+          super.traverse(tree)
+      }
+    }
+
+    /**
+     * @return true, if the tree contained an unsupported await.
+     */
+    private def reportUnsupportedAwait(tree: Tree, whyUnsupported: String): Boolean = {
+      val badAwaits = ListBuffer[Tree]()
+      object traverser extends Traverser {
```

Comment:
Would it make sense to pull these out to the CompilationUnit level? It looks like you're allocating both objects for each node in the traversal.

Reply:
We already do a pre-processing pass to mark the entire AST with `ContainsAwait` / `NoAwait` tree attachments (`markContainsAwait`). These annotations are used to efficiently check for `containsAwait(tree)` above. So we only get here in erroneous code and traverse to find the exact location to report the error, so its not worth optimizing here.